File: q3/JAVA/q3-2.java
import java.util.ArrayList;
import java.util.Stack;

public class Solution {
    public ArrayList<Integer> printListFromTailToHead(ListNode listNode) {
        ArrayList<Integer> result = new ArrayList<Integer>();
        ListNode pNode = listNode;
        // Inverse List
        ListNode preNode = null;
        ListNode tempNode = null;
        while (pNode != null) {
            tempNode = preNode;
            preNode = pNode;
            pNode = pNode.next;
            preNode.next = tempNode;
        }
        listNode = preNode;
        pNode = listNode;
        
        while (pNode != null) {
            result.add(pNode.val);
            pNode = pNode.next;
        }
        return result;
    }
}

File: JZOfferSolution/q03 从尾到头重建链表/C++/q3-2.cpp
class Solution {
public:
    vector<int> printListFromTailToHead(ListNode* head) {
        stack<int> myStack;
        vector<int> result;
        // Reverse List
        ListNode* pNode = head;
        ListNode* preNode = nullptr;
        ListNode* tempNode = nullptr;
        while (pNode != nullptr) {
            tempNode = preNode;
            preNode = pNode;
            pNode = pNode->next;
            preNode->next = tempNode;
        }
        head = preNode;
        // Output
        pNode = head;
        while (pNode != nullptr) {
            result.push_back(pNode->val);
            pNode = pNode->next;
        }
        
        return result;
    }
};
